### PR TITLE
metrics: retry the entire shell builder on failure

### PR DESCRIPTION
**Pull request by paride**
This introduces a new retry mechanism I plan to use on Jenkins jobs for
which there isn't really value in the individual failures. This is the
case of e.g. hourly jobs collecting metrics data over the network.

The new snippet on top of the shell builders causes the script itself to
be re-run a few times if the ERR trap is triggered. While it's ugly to
repeat the same boilerplate, pulling it from the Internet kind of
defeats its objective (= retry if there are network issues). As a
compromise I tried to make the boilerplate short (admittedly at the
cost of readability).

As implemented the retry logic tries a total of 5 times (including the
first one), and waits increasing intervals during the tries, with cubic
backoff, for a total of 1800s of sleep at most.

### CODE REVIEW

**Review by athos-ribeiro — COMMENTED** [review]
Nice work, Paride!

While the commit message says we are limiting the logic to 5 attempts, isn't it actually trying 6 times (5 re-attempts)? This led me to wonder if we could define some variables for the number of retries and backoff factors to make it easier to tweak with those in the future.

Also, although I understand the motivation on how this has been written, it does affect readability, as you mentioned. Do you think adding a longer comment with  a description for the retry function would help our future selves?

Finally, I added an inline comment on the `set -/+ x` logic throughout the script.

File: jenkins-jobs/metrics/metrics.yaml [hunk 1]
```diff
@@ -211,7 +235,17 @@
...
+          set -ux -o pipefail
+
+          ### retry script snippet
+          retry() {{
+              ((!(rc=$?) || ${{_try:=1}}>5)) && exit $rc || sleep "$((8*_try**3))"
+              [[ -n "${{WORKSPACE:-}}" ]] && cd "$WORKSPACE" && find . -delete
+              _try=$((++_try)) exec "$BASH" "${{BASH_SOURCE[-1]}}"
+          }}
+          trap retry ERR
+          ### end of retry snippet
```
[the remark this answers is not in the document]
Comment (paride) [reply jenkins-jobs/metrics/metrics.yaml:248]: Good catch, adding it, thanks!

**Comment by paride** [comment]
Thanks fore reviewing this @athos-ribeiro. You're right, it's 5 *retries*. It used to 5 *tries* (including the first), then I changed the logic a bit and now it's trying 5 times on top on the original try. I still prefer this logic, so I'll amend the commit message.

On making it more readable and maybe configurable: I'll check again if I can come up with a way to add the snippet to many jobs without repeating it. I have a couple of ideas, I need to check if they'll actually work.

I'm making this a WIP.

**Comment by paride** [comment]
This is a prerequisite of the new approach: https://github.com/canonical/server-jenkins-jobs/pull/226.

**Comment by paride** [comment]
This is v2 of the new retry approach, ready for review. Hopefully it's more readable and I like that it doesn't duplicate code, OTOH following yaml anchors is not fun...

This should work well in many cases, however the approach has gotchas:
 - The process environment is not fully reset
 - If more than one shell builder is present, this is only really safe for use in the first one (because of the `find -delete`).

There are better approaches:
 - The Navigator plugin (https://plugins.jenkins.io/naginator/) adds a "retry job" post-build action. We don't have this plugin on our Jenkins instances.
 - Pipeline jobs may allow retrials natively, I have to check. Still I don't think we're going to convert all our jobs to pipeline jobs, so this will still be useful.

TLDR: not the nicest thing on Earth, but hopefully good enough in many cases.

**Comment by paride** [comment]
For the odd `!join` thing look here: https://docs.openstack.org/infra/jenkins-job-builder/definition.html#action-tags

**Review by athos-ribeiro — APPROVED** [review]
Great work, Paride! Thank you :)

**Comment by paride** [comment]
Thanks! I'm merging this and will keep an eye on the metrics jobs. If it works fine there I'll apply this to other jobs.
